### PR TITLE
Yolov5

### PR DESCRIPTION
**Pull request by maxhs2014**
Adding yolov5 model file as detection engine

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,7 +60,8 @@ pip install opencv pycocotools tensorflow
 Run
 
 ```bash
-python clip_object_tracker.py --source data/video/cards.mp4 --url https://detect.roboflow.com/playing-cards-ow27d/1 --api_key ROBOFLOW_API_KEY
+
+ --url https://detect.roboflow.com/playing-cards-ow27d/1 --api_key ROBOFLOW_API_KEY
 ```
```
Comment (Jacobsolawetz) [inline README.md:66]: should we include something in the README with a bash command on how to use the yolov5 engine?

File: clip_object_tracker.py [hunk 2]
```diff
@@ -24,6 +24,12 @@
 from deep_sort.tracker import Tracker
 from tools import generate_clip_detections as gdet
 
+from utils.yolov5 import Yolov5Engine
+
+classes = []
+
```
Comment (Jacobsolawetz) [inline clip_object_tracker.py:30]: should they put their class names here? Maybe we should have class names as a separate file. Better yet, the same data.yaml that they use in custom training.

File: clip_object_tracker.py [hunk 3]
```diff
@@ -85,11 +91,18 @@ def detect(save_img=False):
...
+    half = device != "cpu"
     model, transform = clip.load(model_filename, device=device)
     encoder = gdet.create_box_encoder(model, transform, batch_size=1, device=device)
     # calculate cosine distance metric
     metric = nn_matching.NearestNeighborDistanceMetric(
         "cosine", max_cosine_distance, nn_budget)
+    
+    # load yolov5 model here
+    if opt.detection_engine == "yolov5":
+        yolov5_engine = Yolov5Engine("models/yolov5s.pt", device, opt.classes, opt.confidence, opt.overlap, opt.agnostic_nms, opt.augment, half)
```
Comment (Jacobsolawetz) [inline clip_object_tracker.py:103]: let's make the model path an opt, if not specified we can default to this one

File: README.md (hunk 1 above)
Comment (Jacobsolawetz) [reply README.md:66]: Something here too on which version of YOLOv5 is compatible

File: utils/yolov5.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        pred = self.model(img, augment=self.augment)[0]
+        pred = self.nms(pred)
+        return pred
+
+    def nms(self, pred):
+        out = non_max_suppression(pred, self.conf_thres, self.iou_thres, classes=self.classes, agnostic=self.agnostic_nms)
+        return out
+
+    def get_names(self):
+        return self.model.module.names if hasattr(self.model, 'module') else self.model.names
```
Comment (Jacobsolawetz) [inline utils/yolov5.py:25]: Oh wow, this is cool - did Glenn add the class names into the model definition? If so you can disregard what I was saying ^^

**Review by Jacobsolawetz — COMMENTED** [review]
Overall looks great! I suggested a few revisions

The cars gif looks excellent - it would also be a good idea to try it on a custom dataset (perhaps on cards again) to verify that there aren't any hiccups along the way.

File: clip_object_tracker.py (hunk 2 above)
Comment (maxhs2014) [reply clip_object_tracker.py:30]: no. the class names are loaded form the model right now

File: utils/yolov5.py (hunk 4 above)
Comment (maxhs2014) [reply utils/yolov5.py:25]: Yeah, it loads them from the model

**Pull request merged**